File: packages/less/src/less/visitors/import_sequencer_test.go
package visitors

import (
	"fmt"
	"testing"
)

func TestImportSequencer(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("initializes with empty imports and variableImports arrays", func(t *testing.T) {
			var onSequencerEmpty func()
			sequencer := NewImportSequencer(onSequencerEmpty)

			if len(sequencer.imports) != 0 {
				t.Errorf("Expected imports to be empty, got length %d", len(sequencer.imports))
			}
			if len(sequencer.variableImports) != 0 {
				t.Errorf("Expected variableImports to be empty, got length %d", len(sequencer.variableImports))
			}
		})

		t.Run("stores the onSequencerEmpty callback", func(t *testing.T) {
			called := false
			onSequencerEmpty := func() {
				called = true
			}
			sequencer := NewImportSequencer(onSequencerEmpty)

			if sequencer.onSequencerEmpty == nil {
				t.Error("Expected onSequencerEmpty to be stored")
			}
			// Test that the callback works
			sequencer.onSequencerEmpty()
			if !called {
				t.Error("Expected onSequencerEmpty callback to be called")
			}
		})

		t.Run("initializes currentDepth to 0", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			if sequencer.currentDepth != 0 {
				t.Errorf("Expected currentDepth to be 0, got %d", sequencer.currentDepth)
			}
		})

		t.Run("can be created without onSequencerEmpty callback", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			if sequencer.onSequencerEmpty != nil {
				t.Error("Expected onSequencerEmpty to be nil")
			}
		})
	})

	t.Run("addImport", func(t *testing.T) {
		t.Run("adds an import item to the imports array", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback := func(args ...any) {}

			sequencer.AddImport(callback)

			if len(sequencer.imports) != 1 {
				t.Errorf("Expected imports length to be 1, got %d", len(sequencer.imports))
			}

			importItem := sequencer.imports[0]
			if importItem.callback == nil {
				t.Error("Expected callback to be stored")
			}
			if importItem.args != nil {
				t.Error("Expected args to be nil initially")
			}
			if importItem.isReady {
				t.Error("Expected isReady to be false initially")
			}
		})

		t.Run("returns a trigger function", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback := func(args ...any) {}

			trigger := sequencer.AddImport(callback)

			if trigger == nil {
				t.Error("Expected trigger function to be returned")
			}
		})

		t.Run("trigger function sets isReady to true and stores arguments", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			argsReceived := []any{}
			callback := func(args ...any) {
				argsReceived = args
			}

			trigger := sequencer.AddImport(callback)
			trigger("arg1", "arg2", "arg3")

			if len(argsReceived) != 3 {
				t.Errorf("Expected 3 args, got %d", len(argsReceived))
			}
			if argsReceived[0] != "arg1" ||
				argsReceived[1] != "arg2" ||
				argsReceived[2] != "arg3" {
				t.Error("Expected args to be passed correctly")
			}
		})

		t.Run("trigger function calls TryRun", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback := func(args ...any) {}

			trigger := sequencer.AddImport(callback)
			trigger()

			// If TryRun was called, the import should be processed (removed from queue)
			if len(sequencer.imports) != 0 {
				t.Error("Expected TryRun to process the import")
			}
		})

		t.Run("trigger function works with no arguments", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			argsReceived := []any{}
			callback := func(args ...any) {
				argsReceived = args
			}

			trigger := sequencer.AddImport(callback)
			trigger()

			if len(argsReceived) != 0 {
				t.Errorf("Expected 0 args, got %d", len(argsReceived))
			}
		})

		t.Run("can add multiple imports", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback1 := func(args ...any) {}
			callback2 := func(args ...any) {}

			sequencer.AddImport(callback1)
			sequencer.AddImport(callback2)

			if len(sequencer.imports) != 2 {
				t.Errorf("Expected imports length to be 2, got %d", len(sequencer.imports))
			}
		})
	})

	t.Run("addVariableImport", func(t *testing.T) {
		t.Run("adds a callback to the variableImports array", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback := func() {}

			sequencer.AddVariableImport(callback)

			if len(sequencer.variableImports) != 1 {
				t.Errorf("Expected variableImports length to be 1, got %d", len(sequencer.variableImports))
			}
		})

		t.Run("can add multiple variable imports", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callback1 := func() {}
			callback2 := func() {}

			sequencer.AddVariableImport(callback1)
			sequencer.AddVariableImport(callback2)

			if len(sequencer.variableImports) != 2 {
				t.Errorf("Expected variableImports length to be 2, got %d", len(sequencer.variableImports))
			}
		})
	})

	t.Run("tryRun", func(t *testing.T) {
		t.Run("does not run callback if import is not ready", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			called := false
			callback := func(args ...any) {
				called = true
			}

			sequencer.AddImport(callback)
			sequencer.TryRun()

			if called {
				t.Error("Expected callback not to be called")
			}
			if len(sequencer.imports) != 1 {
				t.Errorf("Expected imports length to be 1, got %d", len(sequencer.imports))
			}
		})

		t.Run("runs callback when import is ready", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			called := false
			var receivedArgs []any
			callback := func(args ...any) {
				called = true
				receivedArgs = args
			}

			trigger := sequencer.AddImport(callback)
			trigger("arg1", "arg2")

			if !called {
				t.Error("Expected callback to be called")
			}
			if len(receivedArgs) != 2 || receivedArgs[0] != "arg1" || receivedArgs[1] != "arg2" {
				t.Error("Expected callback to receive correct arguments")
			}
			if len(sequencer.imports) != 0 {
				t.Errorf("Expected imports to be empty after processing, got length %d", len(sequencer.imports))
			}
		})

		t.Run("processes imports in FIFO order", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}

			callback1 := func(args ...any) {
				callOrder = append(callOrder, "callback1")
			}
			callback2 := func(args ...any) {
				callOrder = append(callOrder, "callback2")
			}
			callback3 := func(args ...any) {
				callOrder = append(callOrder, "callback3")
			}

			trigger1 := sequencer.AddImport(callback1)
			trigger2 := sequencer.AddImport(callback2)
			trigger3 := sequencer.AddImport(callback3)

			// Trigger in reverse order to test FIFO
			trigger3("arg3")
			trigger1("arg1")
			trigger2("arg2")

			expectedOrder := []string{"callback1", "callback2", "callback3"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
		})

		t.Run("processes all ready imports before variable imports", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}

			importCallback := func(args ...any) {
				callOrder = append(callOrder, "import")
			}
			variableCallback := func() {
				callOrder = append(callOrder, "variable")
			}

			sequencer.AddVariableImport(variableCallback)
			trigger := sequencer.AddImport(importCallback)
			trigger()

			expectedOrder := []string{"import", "variable"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
		})

		t.Run("processes variable imports in FIFO order", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}

			callback1 := func() {
				callOrder = append(callOrder, "variable1")
			}
			callback2 := func() {
				callOrder = append(callOrder, "variable2")
			}

			sequencer.AddVariableImport(callback1)
			sequencer.AddVariableImport(callback2)
			sequencer.TryRun()

			expectedOrder := []string{"variable1", "variable2"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
			if len(sequencer.variableImports) != 0 {
				t.Errorf("Expected variableImports to be empty, got length %d", len(sequencer.variableImports))
			}
		})

		t.Run("handles mixed imports and variable imports correctly", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}

			importCallback1 := func(args ...any) {
				callOrder = append(callOrder, "import1")
			}
			importCallback2 := func(args ...any) {
				callOrder = append(callOrder, "import2")
			}
			variableCallback1 := func() {
				callOrder = append(callOrder, "variable1")
			}
			variableCallback2 := func() {
				callOrder = append(callOrder, "variable2")
			}

			trigger1 := sequencer.AddImport(importCallback1)
			sequencer.AddVariableImport(variableCallback1)
			trigger2 := sequencer.AddImport(importCallback2)
			sequencer.AddVariableImport(variableCallback2)

			trigger1()
			trigger2()

			expectedOrder := []string{"import1", "import2", "variable1", "variable2"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
		})

		t.Run("stops processing imports when encountering a non-ready import", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			called1 := false
			called2 := false
			called3 := false

			callback1 := func(args ...any) {
				called1 = true
			}
			callback2 := func(args ...any) {
				called2 = true
			}
			callback3 := func(args ...any) {
				called3 = true
			}

			trigger1 := sequencer.AddImport(callback1)
			sequencer.AddImport(callback2) // This one won't be triggered
			trigger3 := sequencer.AddImport(callback3)

			trigger1()
			trigger3()
			sequencer.TryRun()

			if !called1 {
				t.Error("Expected callback1 to be called")
			}
			if called2 {
				t.Error("Expected callback2 not to be called")
			}
			if called3 {
				t.Error("Expected callback3 not to be called")
			}
			if len(sequencer.imports) != 2 {
				t.Errorf("Expected 2 imports to remain, got %d", len(sequencer.imports))
			}
		})

		t.Run("increments and decrements currentDepth", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)

			if sequencer.currentDepth != 0 {
				t.Errorf("Expected currentDepth to be 0 initially, got %d", sequencer.currentDepth)
			}

			sequencer.TryRun()

			if sequencer.currentDepth != 0 {
				t.Errorf("Expected currentDepth to be 0 after completion, got %d", sequencer.currentDepth)
			}
		})

		t.Run("tracks depth correctly during nested calls", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			depthDuringCallback := -1

			callback := func(args ...any) {
				depthDuringCallback = sequencer.currentDepth
				sequencer.TryRun() // Nested call
				// Depth should still be 1 after nested call returns
				if sequencer.currentDepth != 1 {
					t.Errorf("Expected currentDepth to be 1 after nested call, got %d", sequencer.currentDepth)
				}
			}

			trigger := sequencer.AddImport(callback)
			trigger()

			if depthDuringCallback != 1 {
				t.Errorf("Expected currentDepth to be 1 during callback, got %d", depthDuringCallback)
			}
			if sequencer.currentDepth != 0 {
				t.Errorf("Expected currentDepth to be 0 after all calls, got %d", sequencer.currentDepth)
			}
		})

		t.Run("calls onSequencerEmpty when depth reaches 0 and all imports are processed", func(t *testing.T) {
			called := false
			onSequencerEmpty := func() {
				called = true
			}
			sequencer := NewImportSequencer(onSequencerEmpty)

			callback := func(args ...any) {}
			trigger := sequencer.AddImport(callback)
			trigger()

			if !called {
				t.Error("Expected onSequencerEmpty to be called")
			}
		})

		t.Run("does not call onSequencerEmpty when depth is not 0", func(t *testing.T) {
			callCount := 0
			onSequencerEmpty := func() {
				callCount++
			}
			sequencer := NewImportSequencer(onSequencerEmpty)

			callback := func(args ...any) {
				sequencer.TryRun() // This creates a nested call
			}
			trigger := sequencer.AddImport(callback)
			trigger()

			// onSequencerEmpty should only be called once, after the outermost TryRun completes
			if callCount != 1 {
				t.Errorf("Expected onSequencerEmpty to be called once, got %d times", callCount)
			}
		})

		t.Run("does not call onSequencerEmpty when callback is not provided", func(t *testing.T) {
			sequencer := NewImportSequencer(nil) // No callback provided

			callback := func(args ...any) {}
			trigger := sequencer.AddImport(callback)

			// Should not panic
			defer func() {
				if r := recover(); r != nil {
					t.Errorf("TryRun should not panic when onSequencerEmpty is nil: %v", r)
				}
			}()

			trigger()
		})

		t.Run("handles exceptions in import callbacks gracefully", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)

			errorCallback := func(args ...any) {
				panic("Test error")
			}
			normalCallback := func(args ...any) {}

			sequencer.AddImport(errorCallback)
			sequencer.AddImport(normalCallback)

			// Need to manually set ready state since we're testing exception handling
			sequencer.imports[0].isReady = true
			sequencer.imports[0].args = []any{}
			sequencer.imports[1].isReady = true
			sequencer.imports[1].args = []any{}

			defer func() {
				if r := recover(); r == nil {
					t.Error("Expected TryRun to panic with 'Test error'")
				}
				// currentDepth should be properly decremented even after exception
				if sequencer.currentDepth != 0 {
					t.Errorf("Expected currentDepth to be 0 after exception, got %d", sequencer.currentDepth)
				}
			}()

			sequencer.TryRun()
		})

		t.Run("handles exceptions in variable import callbacks gracefully", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)

			errorCallback := func() {
				panic("Variable import error")
			}

			sequencer.AddVariableImport(errorCallback)

			defer func() {
				if r := recover(); r == nil {
					t.Error("Expected TryRun to panic with 'Variable import error'")
				}
				if sequencer.currentDepth != 0 {
					t.Errorf("Expected currentDepth to be 0 after exception, got %d", sequencer.currentDepth)
				}
			}()

			sequencer.TryRun()
		})

		t.Run("processes imports that become ready during execution", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}
			var trigger2 func(...any)

			callback1 := func(args ...any) {
				callOrder = append(callOrder, "callback1")
				// Make the second import ready during execution of first
				trigger2("arg2")
			}

			callback2 := func(args ...any) {
				callOrder = append(callOrder, "callback2")
			}

			trigger1 := sequencer.AddImport(callback1)
			trigger2 = sequencer.AddImport(callback2)

			trigger1("arg1")

			expectedOrder := []string{"callback1", "callback2"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
		})

		t.Run("handles variable imports that add new imports", func(t *testing.T) {
			sequencer := NewImportSequencer(nil)
			callOrder := []string{}

			variableCallback := func() {
				callOrder = append(callOrder, "variable")
				newImportCallback := func(args ...any) {
					callOrder = append(callOrder, "newImport")
				}
				trigger := sequencer.AddImport(newImportCallback)
				trigger()
			}

			regularCallback := func(args ...any) {
				callOrder = append(callOrder, "regular")
			}

			sequencer.AddVariableImport(variableCallback)
			trigger := sequencer.AddImport(regularCallback)
			trigger()

			expectedOrder := []string{"regular", "variable", "newImport"}
			if len(callOrder) != len(expectedOrder) {
				t.Errorf("Expected %d callbacks, got %d", len(expectedOrder), len(callOrder))
			}
			for i, expected := range expectedOrder {
				if i >= len(callOrder) || callOrder[i] != expected {
					t.Errorf("Expected callback order %v, got %v", expectedOrder, callOrder)
					break
				}
			}
		})
	})

	t.Run("integration scenarios", func(t *testing.T) {
		t.Run("handles complex workflow with multiple import types", func(t *testing.T) {
			called := false
			onSequencerEmpty := func() {
				called = true
			}
			sequencer := NewImportSequencer(onSequencerEmpty)
			results := []string{}

			// Add some variable imports
			sequencer.AddVariableImport(func() { results = append(results, "var1") })
			sequencer.AddVariableImport(func() { results = append(results, "var2") })

			// Add regular imports
			trigger1 := sequencer.AddImport(func(args ...any) { results = append(results, "import1") })
			trigger2 := sequencer.AddImport(func(args ...any) { results = append(results, "import2") })

			// Add more variable imports
			sequencer.AddVariableImport(func() { results = append(results, "var3") })

			// Trigger imports
			trigger2()
			trigger1()

			expectedOrder := []string{"import1", "import2", "var1", "var2", "var3"}
			if len(results) != len(expectedOrder) {
				t.Errorf("Expected %d results, got %d", len(expectedOrder), len(results))
			}
			for i, expected := range expectedOrder {
				if i >= len(results) || results[i] != expected {
					t.Errorf("Expected results %v, got %v", expectedOrder, results)
					break
				}
			}
			if !called {
				t.Error("Expected onSequencerEmpty to be called")
			}
		})

		t.Run("handles recursive import additions", func(t *testing.T) {
			called := false
			onSequencerEmpty := func() {
				called = true
			}
			sequencer := NewImportSequencer(onSequencerEmpty)
			results := []string{}
			recursionCount := 0

			var recursiveCallback func(args ...any)
			recursiveCallback = func(args ...any) {
				results = append(results, fmt.Sprintf("recursive%d", recursionCount))
				recursionCount++

				if recursionCount < 3 {
					trigger := sequencer.AddImport(recursiveCallback)
					trigger()
				}
			}

			trigger := sequencer.AddImport(recursiveCallback)
			trigger()

			expectedOrder := []string{"recursive0", "recursive1", "recursive2"}
			if len(results) != len(expectedOrder) {
				t.Errorf("Expected %d results, got %d", len(expectedOrder), len(results))
			}
			for i, expected := range expectedOrder {
				if i >= len(results) || results[i] != expected {
					t.Errorf("Expected results %v, got %v", expectedOrder, results)
					break
				}
			}
			if !called {
				t.Error("Expected onSequencerEmpty to be called")
			}
		})

		t.Run("maintains correct state when no imports are added", func(t *testing.T) {
			called := false
			onSequencerEmpty := func() {
				called = true
			}
			sequencer := NewImportSequencer(onSequencerEmpty)

			sequencer.TryRun()

			if len(sequencer.imports) != 0 {
				t.Errorf("Expected imports to be empty, got length %d", len(sequencer.imports))
			}
			if len(sequencer.variableImports) != 0 {
				t.Errorf("Expected variableImports to be empty, got length %d", len(sequencer.variableImports))
			}
			if sequencer.currentDepth != 0 {
				t.Errorf("Expected currentDepth to be 0, got %d", sequencer.currentDepth)
			}
			if !called {
				t.Error("Expected onSequencerEmpty to be called")
			}
		})
	})
}

File: packages/less/src/less/visitors/import-sequencer.test.js
import { describe, it, expect, vi, beforeEach } from 'vitest';
import ImportSequencer from './import-sequencer.js';

describe('ImportSequencer', () => {
    let sequencer;
    let onSequencerEmpty;

    beforeEach(() => {
        onSequencerEmpty = vi.fn();
        sequencer = new ImportSequencer(onSequencerEmpty);
    });

    describe('constructor', () => {
        it('initializes with empty imports and variableImports arrays', () => {
            expect(sequencer.imports).toEqual([]);
            expect(sequencer.variableImports).toEqual([]);
        });

        it('stores the onSequencerEmpty callback', () => {
            expect(sequencer._onSequencerEmpty).toBe(onSequencerEmpty);
        });

        it('initializes _currentDepth to 0', () => {
            expect(sequencer._currentDepth).toBe(0);
        });

        it('can be created without onSequencerEmpty callback', () => {
            const seq = new ImportSequencer();
            expect(seq._onSequencerEmpty).toBeUndefined();
        });
    });

    describe('addImport', () => {
        it('adds an import item to the imports array', () => {
            const callback = vi.fn();
            sequencer.addImport(callback);

            expect(sequencer.imports).toHaveLength(1);
            expect(sequencer.imports[0]).toEqual({
                callback,
                args: null,
                isReady: false
            });
        });

        it('returns a trigger function', () => {
            const callback = vi.fn();
            const trigger = sequencer.addImport(callback);

            expect(typeof trigger).toBe('function');
        });

        it('trigger function sets isReady to true and stores arguments', () => {
            const callback = vi.fn();
            const trigger = sequencer.addImport(callback);

            // Mock tryRun to prevent it from removing the import
            const originalTryRun = sequencer.tryRun;
            sequencer.tryRun = vi.fn();

            trigger('arg1', 'arg2', 'arg3');

            expect(sequencer.imports[0].isReady).toBe(true);
            expect(sequencer.imports[0].args).toEqual(['arg1', 'arg2', 'arg3']);

            // Restore original tryRun
            sequencer.tryRun = originalTryRun;
        });

        it('trigger function calls tryRun', () => {
            const callback = vi.fn();
            const tryRunSpy = vi.spyOn(sequencer, 'tryRun');
            const trigger = sequencer.addImport(callback);

            trigger();

            expect(tryRunSpy).toHaveBeenCalled();
        });

        it('trigger function works with no arguments', () => {
            const callback = vi.fn();
            const trigger = sequencer.addImport(callback);

            // Mock tryRun to prevent it from removing the import
            const originalTryRun = sequencer.tryRun;
            sequencer.tryRun = vi.fn();

            trigger();

            expect(sequencer.imports[0].args).toEqual([]);
            expect(sequencer.imports[0].isReady).toBe(true);

            // Restore original tryRun
            sequencer.tryRun = originalTryRun;
        });

        it('can add multiple imports', () => {
            const callback1 = vi.fn();
            const callback2 = vi.fn();

            sequencer.addImport(callback1);
            sequencer.addImport(callback2);

            expect(sequencer.imports).toHaveLength(2);
            expect(sequencer.imports[0].callback).toBe(callback1);
            expect(sequencer.imports[1].callback).toBe(callback2);
        });
    });

    describe('addVariableImport', () => {
        it('adds a callback to the variableImports array', () => {
            const callback = vi.fn();
            sequencer.addVariableImport(callback);

            expect(sequencer.variableImports).toHaveLength(1);
            expect(sequencer.variableImports[0]).toBe(callback);
        });

        it('can add multiple variable imports', () => {
            const callback1 = vi.fn();
            const callback2 = vi.fn();

            sequencer.addVariableImport(callback1);
            sequencer.addVariableImport(callback2);

            expect(sequencer.variableImports).toHaveLength(2);
            expect(sequencer.variableImports[0]).toBe(callback1);
            expect(sequencer.variableImports[1]).toBe(callback2);
        });
    });

    describe('tryRun', () => {
        it('does not run callback if import is not ready', () => {
            const callback = vi.fn();
            sequencer.addImport(callback);

            sequencer.tryRun();

            expect(callback).not.toHaveBeenCalled();
            expect(sequencer.imports).toHaveLength(1);
        });

        it('runs callback when import is ready', () => {
            const callback = vi.fn();
            const trigger = sequencer.addImport(callback);

            trigger('arg1', 'arg2');
            // tryRun is called automatically by trigger, but let's call it explicitly for clarity

            expect(callback).toHaveBeenCalledWith('arg1', 'arg2');
            expect(sequencer.imports).toHaveLength(0);
        });

        it('processes imports in FIFO order', () => {
            const callback1 = vi.fn();
            const callback2 = vi.fn();
            const callback3 = vi.fn();
            const callOrder = [];

            callback1.mockImplementation(() => callOrder.push('callback1'));
            callback2.mockImplementation(() => callOrder.push('callback2'));
            callback3.mockImplementation(() => callOrder.push('callback3'));

            const trigger1 = sequencer.addImport(callback1);
            const trigger2 = sequencer.addImport(callback2);
            const trigger3 = sequencer.addImport(callback3);

            // Trigger in reverse order to test FIFO
            trigger3('arg3');
            trigger1('arg1');
            trigger2('arg2');

            expect(callOrder).toEqual(['callback1', 'callback2', 'callback3']);
        });

        it('processes all ready imports before variable imports', () => {
            const importCallback = vi.fn();
            const variableCallback = vi.fn();
            const callOrder = [];

            importCallback.mockImplementation(() => callOrder.push('import'));
            variableCallback.mockImplementation(() =>
                callOrder.push('variable')
            );

            sequencer.addVariableImport(variableCallback);
            const trigger = sequencer.addImport(importCallback);

            trigger();

            expect(callOrder).toEqual(['import', 'variable']);
        });

        it('processes variable imports in FIFO order', () => {
            const callback1 = vi.fn();
            const callback2 = vi.fn();
            const callOrder = [];

            callback1.mockImplementation(() => callOrder.push('variable1'));
            callback2.mockImplementation(() => callOrder.push('variable2'));

            sequencer.addVariableImport(callback1);
            sequencer.addVariableImport(callback2);

            sequencer.tryRun();

            expect(callOrder).toEqual(['variable1', 'variable2']);
            expect(sequencer.variableImports).toHaveLength(0);
        });

        it('handles mixed imports and variable imports correctly', () => {
            const importCallback1 = vi.fn();
            const importCallback2 = vi.fn();
            const variableCallback1 = vi.fn();
            const variableCallback2 = vi.fn();
            const callOrder = [];

            importCallback1.mockImplementation(() => callOrder.push('import1'));
            importCallback2.mockImplementation(() => callOrder.push('import2'));
            variableCallback1.mockImplementation(() =>
                callOrder.push('variable1')
            );
            variableCallback2.mockImplementation(() =>
                callOrder.push('variable2')
            );

            const trigger1 = sequencer.addImport(importCallback1);
            sequencer.addVariableImport(variableCallback1);
            const trigger2 = sequencer.addImport(importCallback2);
            sequencer.addVariableImport(variableCallback2);

            trigger1();
            trigger2();

            expect(callOrder).toEqual([
                'import1',
                'import2',
                'variable1',
                'variable2'
            ]);
        });

        it('stops processing imports when encountering a non-ready import', () => {
            const callback1 = vi.fn();
            const callback2 = vi.fn();
            const callback3 = vi.fn();

            const trigger1 = sequencer.addImport(callback1);
            sequencer.addImport(callback2); // This one won't be triggered
            const trigger3 = sequencer.addImport(callback3);

            trigger1();
            trigger3();

            sequencer.tryRun();

            expect(callback1).toHaveBeenCalled();
            expect(callback2).not.toHaveBeenCalled();
            expect(callback3).not.toHaveBeenCalled();
            expect(sequencer.imports).toHaveLength(2); // callback2 and callback3 remain
        });

        it('increments and decrements _currentDepth', () => {
            expect(sequencer._currentDepth).toBe(0);

            sequencer.tryRun();

            expect(sequencer._currentDepth).toBe(0); // Should be back to 0 after completion
        });

        it('tracks depth correctly during nested calls', () => {
            const callback = vi.fn(() => {
                expect(sequencer._currentDepth).toBe(1);
                sequencer.tryRun(); // Nested call
                expect(sequencer._currentDepth).toBe(1); // Should still be 1 after nested call returns
            });

            const trigger = sequencer.addImport(callback);
            trigger();

            expect(sequencer._currentDepth).toBe(0);
        });

        it('calls onSequencerEmpty when depth reaches 0 and all imports are processed', () => {
            const callback = vi.fn();
            const trigger = sequencer.addImport(callback);

            trigger();

            expect(onSequencerEmpty).toHaveBeenCalled();
        });

        it('does not call onSequencerEmpty when depth is not 0', () => {
            const callback = vi.fn(() => {
                sequencer.tryRun(); // This creates a nested call
            });
            const trigger = sequencer.addImport(callback);

            trigger();

            // onSequencerEmpty should only be called once, after the outermost tryRun completes
            expect(onSequencerEmpty).toHaveBeenCalledTimes(1);
        });

        it('does not call onSequencerEmpty when callback is not provided', () => {
            const seq = new ImportSequencer(); // No callback provided
            const callback = vi.fn();
            const trigger = seq.addImport(callback);

            trigger();

            // Should not throw an error
            expect(() => seq.tryRun()).not.toThrow();
        });

        it('handles exceptions in import callbacks gracefully', () => {
            const errorCallback = vi.fn(() => {
                throw new Error('Test error');
            });
            const normalCallback = vi.fn();

            sequencer.addImport(errorCallback);
            sequencer.addImport(normalCallback);

            // Need to manually set ready state since we're testing exception handling
            sequencer.imports[0].isReady = true;
            sequencer.imports[0].args = [];
            sequencer.imports[1].isReady = true;
            sequencer.imports[1].args = [];

            expect(() => sequencer.tryRun()).toThrow('Test error');
            // _currentDepth should be properly decremented even after exception
            expect(sequencer._currentDepth).toBe(0);
        });

        it('handles exceptions in variable import callbacks gracefully', () => {
            const errorCallback = vi.fn(() => {
                throw new Error('Variable import error');
            });

            sequencer.addVariableImport(errorCallback);

            expect(() => sequencer.tryRun()).toThrow('Variable import error');
            expect(sequencer._currentDepth).toBe(0);
        });

        it('processes imports that become ready during execution', () => {
            const callOrder = [];
            let trigger2;

            const callback1 = vi.fn(() => {
                callOrder.push('callback1');
                // Make the second import ready during execution of first
                trigger2('arg2');
            });

            const callback2 = vi.fn(() => {
                callOrder.push('callback2');
            });

            const trigger1 = sequencer.addImport(callback1);
            trigger2 = sequencer.addImport(callback2);

            trigger1('arg1');

            expect(callOrder).toEqual(['callback1', 'callback2']);
        });

        it('handles variable imports that add new imports', () => {
            const callOrder = [];

            const variableCallback = vi.fn(() => {
                callOrder.push('variable');
                const newImportCallback = vi.fn(() =>
                    callOrder.push('newImport')
                );
                const trigger = sequencer.addImport(newImportCallback);
                trigger();
            });

            const regularCallback = vi.fn(() => callOrder.push('regular'));

            sequencer.addVariableImport(variableCallback);
            const trigger = sequencer.addImport(regularCallback);
            trigger();

            expect(callOrder).toEqual(['regular', 'variable', 'newImport']);
        });
    });

    describe('integration scenarios', () => {
        it('handles complex workflow with multiple import types', () => {
            const results = [];

            // Add some variable imports
            sequencer.addVariableImport(() => results.push('var1'));
            sequencer.addVariableImport(() => results.push('var2'));

            // Add regular imports
            const trigger1 = sequencer.addImport(() => results.push('import1'));
            const trigger2 = sequencer.addImport(() => results.push('import2'));

            // Add more variable imports
            sequencer.addVariableImport(() => results.push('var3'));

            // Trigger imports
            trigger2();
            trigger1();

            expect(results).toEqual([
                'import1',
                'import2',
                'var1',
                'var2',
                'var3'
            ]);
            expect(onSequencerEmpty).toHaveBeenCalled();
        });

        it('handles recursive import additions', () => {
            const results = [];
            let recursionCount = 0;

            const recursiveCallback = vi.fn(() => {
                results.push(`recursive${recursionCount}`);
                recursionCount++;

                if (recursionCount < 3) {
                    const trigger = sequencer.addImport(recursiveCallback);
                    trigger();
                }
            });

            const trigger = sequencer.addImport(recursiveCallback);
            trigger();

            expect(results).toEqual(['recursive0', 'recursive1', 'recursive2']);
            expect(onSequencerEmpty).toHaveBeenCalled();
        });

        it('maintains correct state when no imports are added', () => {
            sequencer.tryRun();

            expect(sequencer.imports).toEqual([]);
            expect(sequencer.variableImports).toEqual([]);
            expect(sequencer._currentDepth).toBe(0);
            expect(onSequencerEmpty).toHaveBeenCalled();
        });
    });
});
